bugsnag-android-core/src/androidTest/java/com/bugsnag/android/BugsnagExceptionTest.kt

```diff
@@ -0,0 +1,82 @@
+import com.bugsnag.android.BugsnagTestUtils.streamableToJsonArray
+import org.json.JSONArray
+import org.json.JSONObject
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+
+class BugsnagExceptionTest {
+
+    private val config = generateConfiguration()
+
+    @Test
+    fun stacktraceConstructorSerialisation() {
+        val trace = arrayOf(StackTraceElement("Foo", "bar", "Foo.kt", 1))
+        val exc = jsonObjectFromException(BugsnagException("MyClass", "Custom message", trace))
+        assertEquals("MyClass", exc.get("errorClass"))
+        assertEquals("Custom message", exc.get("message"))
+        assertEquals("android", exc.get("type"))
+
+        val stacktrace = exc.get("stacktrace") as JSONArray
+        assertTrue(stacktrace.length() > 0)
+    }
+
+    @Test
+    fun throwableConstructorSerialisation() {
+        val exc = jsonObjectFromException(BugsnagException(RuntimeException("oops")))
+        assertEquals("java.lang.RuntimeException", exc.get("errorClass"))
+        assertEquals("oops", exc.get("message"))
+        assertEquals("android", exc.get("type"))
+
+        val stacktrace = exc.get("stacktrace") as JSONArray
+        assertTrue(stacktrace.length() > 0)
+    }
+
+    @Test
+    fun overrideName() {
+        val bugsnagException = BugsnagException(RuntimeException("oops"))
+        bugsnagException.name = "FatalNetworkError"
+        val exc = jsonObjectFromException(bugsnagException)
+        assertEquals("FatalNetworkError", exc.get("errorClass"))
+    }
+
+    @Test
+    fun overrideMessage() {
+        val bugsnagException = BugsnagException(RuntimeException("oops"))
+        bugsnagException.setMessage("User not found")
+        val exc = jsonObjectFromException(bugsnagException)
+        assertEquals("User not found", exc.get("message"))
+    }
+
+    @Test
+    fun overrideType() {
+        val bugsnagException = BugsnagException(RuntimeException("oops"))
+        bugsnagException.type = "browserjs"
+        val exc = jsonObjectFromException(bugsnagException)
+        assertEquals("browserjs", exc.get("type"))
+    }
+
+    @Test
+    fun nestedBugsnagException() {
+        val nestedThrowable = BugsnagException(BugsnagException(RuntimeException("oops")))
+        val exc = jsonObjectFromException(nestedThrowable)
+        assertEquals("java.lang.RuntimeException", exc.get("errorClass"))
+        assertEquals("oops", exc.get("message"))
+        assertEquals("android", exc.get("type"))
+
+        val stacktrace = exc.get("stacktrace") as JSONArray
+        assertTrue(stacktrace.length() > 0)
+    }
+
+    private fun jsonObjectFromException(bugsnagException: BugsnagException): JSONObject {
+        val exceptions = Exceptions(config, bugsnagException)
+        val json = streamableToJsonArray(exceptions)
+        val exc = json.get(0) as JSONObject
+        return exc
+    }
+
+    // TODO convert JavaScriptException to BugsnagException and create Stacktrace
```

Comment:
Any reason this can't be done now? Is it lengthy to do?

Reply:
I added this comment early on in the implementation of this fix and evidently forgot to remove it. It turns out that `JavaScriptException` adds in an additional `columnNumber` field in its serialisation code, so it can't use `BugsnagException` unless we update the implementation of that class to handle this. I also felt that converting `JavaScriptException` to `BugsnagException` would make this changeset larger than it needed to be to fix the issue, and risk introducing some other issue.